docs: improve customization chapter (#1057)
acdh-oeaw/apis-core-rdf: docs/source/customization.rst

Customization
=============

APIS is designed to be easily customizable. This section describes how you can

> customize the views and templates of the generic views that are shipped with

Reviewer: ```suggestion
customize the views and the templates of the views that are shipped with
```

> APIS by overwriting the classes and methods that are used by the generic views.

Reviewer: ```suggestion
APIS by injecting classes and methods that are then automatically used by the generic views.
```

its models. This means that you can use the generic views for the historical 
models of your own ontology. E.g. if you have a model ``Person`` in your 

> `apis_ontology.models` module, you can use the generic views for it. 

Reviewer: ```suggestion
``apis_ontology.models`` module, you can use the generic views for it. 
```

> `/apis/apis_ontology.historicalperson/` 

Reviewer: ```suggestion
``/apis/apis_ontology.historicalperson/`` 
```

will be the URL for the list view of the historical model. 

> `/apis/api/apis_ontology.historicalperson/` for the API view.

Reviewer: ```suggestion
``/apis/api/apis_ontology.historicalperson/`` for the API view.
```

all the parent models following the full inheritance chain. So if all your models
inherit from ``MyAbstractModel``, you can for example create an override table
for all your models by creating a ``myproject.tables.MyAbstractModelTable``.

> Extending autocomplete results

Reviewer: This duplicates in part the part from line 115-120, so it might make sense to delete those lines

Author: yes, but it contains actual example code....trying to move it in the upper section

Reviewer: Yes, I meant removing the older version or merging it together

------------------------------

The autocomplete endpoint :py:class:`apis_core.generic.views.Autocomplete` can
be extended to provide additional autocomplete results, that are not

> based on the default Django querysets. The autocomplete view always refers to a

Reviewer: > The autocomplete view always refers to a Django model

It took me a while to figure out what that refers to. Given that *all* the views in the `generic.views` model use a model as a basis, I think this sentence can be removed (or moved to the top).

Author: ok, removed it

Django model. To extend the results for a specific model, you have to create an

> `ExternalAutocomplete` class, that is named after the model and resides in the

Reviewer: ```suggestion
``ExternalAutocomplete`` class, that is named after the model and resides in the
```

> `querysets` module in the same app. So if you have an app called `myapp` with a

Reviewer: ```suggestion
``querysets`` module in the same app. So if you have an app called ``myapp`` with a
```

> `models.py`

Reviewer: ```suggestion
``models.py``
```

.. code-block:: python

   class Person(models.Model):
        name = models.CharField(max_length=255)

> then the respective autocomplete class should reside in `myapp.querysets` and

Reviewer: ```suggestion
then the respective autocomplete class should reside in ``myapp.querysets`` and
```

> has to be called `PersonExternalAutocomplete`.

Reviewer: ```suggestion
has to be called ``PersonExternalAutocomplete``.
```

.. code-block:: python

    class PersonExternalAutocomplete:
        def extract_results(data):
            ... do something with the data
            return data

        def get_results(self, q):
            with urllib.request.urlopen(f"https://some.uri.tld/search?q={q}") as f:
                data = extract_results(json.loads(f.read()))
                return results
            return {}

> The class has to have a `get_results` method that receives a query as the first

Reviewer: ```suggestion
The class has to have a ``get_results`` method that receives a query as the first
```

> parameter and returns a result in the format, the `django-autocomplete-light`

Reviewer: ```suggestion
parameter and returns a result in the format, the `django-autocomplete-light <https://django-autocomplete-light.readthedocs.io/>`_
```

module uses- this is a dict with the keys "id", "text" and "selected_text".

Importing data from external resources
--------------------------------------

APIS provides the structure for easily importing data from external resources.

> One main component for this are `Importer` classes. They always belong to a

Reviewer: ```suggestion
One main component for this are ``Importer`` classes. They always belong to a
```

> Django Model, reside in the same app as the Djang model in the `importers`

Reviewer: ```suggestion
Django model, reside in the same app as the Django model in the ``importers``
```

> module and are named after the Django Model. So if you have an app called

Reviewer: ```suggestion
module and are named after the Django model. So if you have an app called
```

> `myapp` with a `models.py`

Reviewer: ```suggestion
``myapp`` with a ``models.py``
```

.. code-block:: python

   class Person(models.Model):
        name = models.CharField(max_length=255)

> then the respective importer should reside in `myapp.importers` and has to be

Reviewer: ```suggestion
then the respective importer should reside in ``myapp.importers`` and has to be
```

> called `PersonImporter`.

Reviewer: ```suggestion
called ``PersonImporter``.
```

> An importer takes two arguments to instantiate: an `uri` and a `model`. The

Reviewer: ```suggestion
An importer takes two arguments to instantiate: an ``uri`` and a ``model``. The
```

importers task is then to create a model instance from this URI, usually by
fetching data from the URI, parsing it and extracting the needed fields.

> The instance should then be returned by the `create_instance` method of the

Reviewer: ```suggestion
The instance should then be returned by the ``create_instance`` method of the
```

> importer. There is :py:class:`apis_core.generic.importers.GenericImporter`

Reviewer: ```suggestion
importer. There is :py:class:`apis_core.generic.importers.GenericModelImporter`
```

> which you can inherit from.

Reviewer: ```suggestion
which you can inherit from. It is used by default of no other importer is defined for the model and it tries to do the right thing out of the box: it first looks if there is an RDF configuration for the URI and if that fails tries to parse the URI response as json.
```

To use this logic in forms, there is
:py:class:`apis_core.generic.forms.fields.ModelImportChoiceField` which is

> based on `django.forms.ModelChoiceField`. It checks if the passed value starts

Reviewer: ```suggestion
based on `django.forms.ModelChoiceField <https://docs.djangoproject.com/en/5.0/ref/forms/fields/#modelchoicefield>`_. It checks if the passed value starts
```

> with `http` and if so, it uses the importer that fits the model and uses it to

Reviewer: ```suggestion
with ``http`` and if so, it uses the importer that fits the model and uses it to
```
